File: integration_tests/dart_client_tests/test/scalars_test.dart
import 'package:test/test.dart';
import 'package:graphlink_client_integration_tests/generated/client/graph_link_client.dart';
import 'fixtures.dart';
import 'mock_adapter.dart';

void main() {
  late MockAdapter adapter;
  late GraphLinkClient client;

  setUp(() {
    adapter = MockAdapter();
    client = GraphLinkClient(adapter: adapter.call, wsAdapter: MockWebSocketAdapter());
  });

  group('non-nullable scalar fields', () {
    setUp(() => adapter.registerData(
          'getAllScalars',
          {'getAllScalars': kAllScalarsNullJson},
        ));

    test('id deserializes to String', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-1');
      expect(res.getAllScalars.id, isA<String>());
      expect(res.getAllScalars.id, equals('scalar-1'));
    });

    test('strVal deserializes to String', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-1');
      expect(res.getAllScalars.strVal, isA<String>());
      expect(res.getAllScalars.strVal, equals('hello world'));
    });

    test('intVal deserializes to int', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-1');
      expect(res.getAllScalars.intVal, isA<int>());
      expect(res.getAllScalars.intVal, equals(42));
    });

    test('floatVal deserializes to double', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-1');
      expect(res.getAllScalars.floatVal, isA<double>());
      expect(res.getAllScalars.floatVal, closeTo(3.14, 0.001));
    });

    test('boolVal deserializes to bool', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-1');
      expect(res.getAllScalars.boolVal, isA<bool>());
      expect(res.getAllScalars.boolVal, isTrue);
    });
  });

  group('nullable scalar fields — null in response', () {
    setUp(() => adapter.registerData(
          'getAllScalars',
          {'getAllScalars': kAllScalarsNullJson},
        ));

    test('nullableStr is null', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-1');
      expect(res.getAllScalars.nullableStr, isNull);
    });

    test('nullableInt is null', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-1');
      expect(res.getAllScalars.nullableInt, isNull);
    });

    test('nullableFloat is null', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-1');
      expect(res.getAllScalars.nullableFloat, isNull);
    });

    test('nullableBool is null', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-1');
      expect(res.getAllScalars.nullableBool, isNull);
    });

    test('nullableId is null', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-1');
      expect(res.getAllScalars.nullableId, isNull);
    });
  });

  group('nullable scalar fields — values present', () {
    setUp(() => adapter.registerData(
          'getAllScalars',
          {'getAllScalars': kAllScalarsPresentJson},
        ));

    test('nullableStr has value', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-2');
      expect(res.getAllScalars.nullableStr, equals('present'));
    });

    test('nullableInt has value', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-2');
      expect(res.getAllScalars.nullableInt, equals(99));
    });

    test('nullableFloat has value', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-2');
      expect(res.getAllScalars.nullableFloat, closeTo(2.718, 0.001));
    });

    test('nullableBool is false (not null)', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-2');
      expect(res.getAllScalars.nullableBool, isFalse);
    });

    test('negative int deserializes correctly', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-2');
      expect(res.getAllScalars.intVal, equals(-1));
    });

    test('negative float deserializes correctly', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-2');
      expect(res.getAllScalars.floatVal, closeTo(-0.5, 0.001));
    });

    test('bool false is not treated as null', () async {
      final res = await client.queries.getAllScalars(id: 'scalar-2');
      expect(res.getAllScalars.boolVal, isFalse);
      expect(res.getAllScalars.boolVal, isNotNull);
    });
  });

  group('outgoing request shape', () {
    setUp(() => adapter.registerData(
          'getAllScalars',
          {'getAllScalars': kAllScalarsNullJson},
        ));

    test('id argument is sent as string in variables', () async {
      await client.queries.getAllScalars(id: 'scalar-1');
      expect(adapter.lastCall!.variables['id'], equals('scalar-1'));
    });

    test('operation name is getAllScalars', () async {
      await client.queries.getAllScalars(id: 'scalar-1');
      expect(adapter.lastCall!.operationName, equals('getAllScalars'));
    });
  });
}

File: integration_tests/typescript_client_tests/test/scalars.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { GraphLinkClient } from '../lib/generated/client/graph-link-client.js';
import { MockAdapter, MockWsAdapter } from './mock-adapter.ts';
import { kAllScalarsNullJson, kAllScalarsPresentJson } from './fixtures.ts';

let adapter: MockAdapter;
let client: GraphLinkClient;

beforeEach(() => {
  adapter = new MockAdapter();
  client = new GraphLinkClient(adapter.call, new MockWsAdapter());
});

describe('getAllScalars — all nullables null', () => {
  beforeEach(() => {
    adapter.registerData('getAllScalars', { getAllScalars: kAllScalarsNullJson });
  });

  it('id', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-1' });
    expect(res.getAllScalars.id).toBe('scalar-1');
  });

  it('strVal', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-1' });
    expect(res.getAllScalars.strVal).toBe('hello world');
  });

  it('intVal', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-1' });
    expect(res.getAllScalars.intVal).toBe(42);
  });

  it('floatVal', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-1' });
    expect(res.getAllScalars.floatVal).toBeCloseTo(3.14);
  });

  it('boolVal', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-1' });
    expect(res.getAllScalars.boolVal).toBe(true);
  });

  it('nullableStr is null', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-1' });
    expect(res.getAllScalars.nullableStr).toBeNull();
  });

  it('nullableInt is null', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-1' });
    expect(res.getAllScalars.nullableInt).toBeNull();
  });

  it('nullableFloat is null', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-1' });
    expect(res.getAllScalars.nullableFloat).toBeNull();
  });

  it('nullableBool is null', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-1' });
    expect(res.getAllScalars.nullableBool).toBeNull();
  });

  it('nullableId is null', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-1' });
    expect(res.getAllScalars.nullableId).toBeNull();
  });
});

describe('getAllScalars — all nullables present', () => {
  beforeEach(() => {
    adapter.registerData('getAllScalars', { getAllScalars: kAllScalarsPresentJson });
  });

  it('negative intVal', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-2' });
    expect(res.getAllScalars.intVal).toBe(-1);
  });

  it('negative floatVal', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-2' });
    expect(res.getAllScalars.floatVal).toBeCloseTo(-0.5);
  });

  it('boolVal false', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-2' });
    expect(res.getAllScalars.boolVal).toBe(false);
  });

  it('nullableStr present', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-2' });
    expect(res.getAllScalars.nullableStr).toBe('present');
  });

  it('nullableInt present', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-2' });
    expect(res.getAllScalars.nullableInt).toBe(99);
  });

  it('nullableFloat present', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-2' });
    expect(res.getAllScalars.nullableFloat).toBeCloseTo(2.718);
  });

  it('nullableBool false (present)', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-2' });
    expect(res.getAllScalars.nullableBool).toBe(false);
  });

  it('nullableId present', async () => {
    const res = await client.queries.getAllScalars({ id: 'scalar-2' });
    expect(res.getAllScalars.nullableId).toBe('nid-1');
  });
});

describe('outgoing variables', () => {
  beforeEach(() => {
    adapter.registerData('getAllScalars', { getAllScalars: kAllScalarsNullJson });
  });

  it('sends the id variable', async () => {
    await client.queries.getAllScalars({ id: 'scalar-99' });
    expect(adapter.lastCall?.variables['id']).toBe('scalar-99');
  });

  it('records exactly one call', async () => {
    await client.queries.getAllScalars({ id: 'scalar-1' });
    expect(adapter.callCount).toBe(1);
  });
});
